javalib/src/main/scala/java/math/BigDecimal.scala

```diff
@@ -1727,11 +1727,24 @@ class BigDecimal() extends Number with Comparable[BigDecimal] {
    *          number don't fit in the primitive type
    */
   private def valueExact(bitLengthOfType: Int): Long = {
-    val bigInteger = toBigIntegerExact()
-    if (bigInteger.bitLength() < bitLengthOfType)
-      bigInteger.longValue()
-    else
+    // Fast path to avoid some large BigInteger creations by toBigIntegerExact
+    val length10 =_scale - _precision
+    if (length10 < 0) {
+      // It is not an integral number
+      throw new ArithmeticException("Rounding necessary")
+    } else if (length10 > bitLengthOfType * 3) {
+      /* length10 is the digit length in base 10 of the integral
+       * number which is less or equal to it's length in octal representation
+       * (i.e. bitLengthOfType * 3)
+       */
       throw new ArithmeticException("Rounding necessary")
```

Comment:
I don't understand this part, and hence I am not convinced it works.

Why not simply use `length10 > 19`, which is `Long.MaxValue.toString.length`, i.e., the largest number of base 10 digits that a `Long` can ever have?